crates/synthesis/src/discrete_basis/solovay_kitaev.rs

```diff
@@ -0,0 +1,370 @@
+        let circuit = output.to_circuit(Some((&matrix_u2, phase)))?;
+        Ok(circuit)
+    }
+
+    /// Run a recursion step for a gate sequence, given a recursion degree.
+    ///
+    /// If the degree is 0 (recursion root), return the closest approximation in the set of
+    /// basic approximations. Otherwise, decompose the difference between the approximation
+    /// and the target unitary as balanced group commutator, and recurse on each element.
+    fn recurse(&self, matrix_so3: &Matrix3<f64>, degree: usize) -> GateSequence {
+        // Recursion root: return the best approximation in the precomputed set.
+        if degree == 0 {
+            if self.do_checks {
+                math::assert_so3("Recursion root", matrix_so3);
+            }
+            let basic_approximation = self
+                .basic_approximations
+                .query(matrix_so3)
+                .expect("No basic approximation in root found")
+                .clone();
+            return basic_approximation;
+        }
+
+        // Find a basic approximation of the target sequence...
+        let u_n1 = self.recurse(matrix_so3, degree - 1);
+
+        // ... and then improve the delta in between that approximation and the target.
+        let delta = matrix_so3 * u_n1.matrix_so3.transpose();
+        let (matrix_vn, matrix_wn) = group_commutator_decomposition(&delta, self.do_checks);
+
+        // Recurse on the group commutator elements.
+        let v_n1 = self.recurse(&matrix_vn, degree - 1);
+        let w_n1 = self.recurse(&matrix_wn, degree - 1);
+
+        v_n1.dot(&w_n1)
+            .dot(&v_n1.adjoint())
+            .dot(&w_n1.adjoint())
+            .dot(&u_n1)
+    }
+
+    /// Store the basic approximations into a file.
+    fn save(&self, filename: &str) -> ::std::io::Result<()> {
+        self.basic_approximations.save(filename)
+    }
+
+    /// Load basic approximation from a file to instantiate this class.
+    fn from_basic_approximations(filename: &str, do_checks: bool) -> ::std::io::Result<Self> {
+        let basic_approximations = BasicApproximations::load(filename)?;
+        Ok(Self {
+            basic_approximations,
+            do_checks,
+        })
+    }
+}
+
+#[pymethods]
+impl SolovayKitaevSynthesis {
+    /// Args:
+    ///     basis_gates (list[Gate] | None): A list of discrete (i.e., non-parameterized) standard
+    ///         gates. Defaults to ``[H, T, Tdg]``.
+    ///     depth (int): The number of basis gate combinations to consider in the basis set. This
+    ///         determines how fast (and if) the algorithm converges and should be chosen
+    ///         sufficiently high. Defaults to 12,
+    ///     tol (float | None): A tolerance determining the granularity of the basic approximations.
+    ///         Any sequence whose SO(3) representation is withing :math:`\sqrt{\texttt{tol}}` of
+    ///         an existing point, will be discarded. Defaults to ``1e-10``.
```

Comment:
I don't see in the code where the default of tol is 1e-10.

Reply:
there are several constants for tol that are defined in the code: sometime it's  1e-10 and sometimes 1e-5.

Reply:
Oh it should say `1e-12`, I'll update.

This one is the only user-exposed `tol`, the other two usages are only internal, local variables. I've renamed/removed their definitions to be less confusing.